sound/soc/sof/intel/hda.c

```diff
@@ -1039,13 +1039,72 @@ static int hda_generic_machine_select(struct snd_sof_dev *sdev)
 #endif
 
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_SOUNDWIRE)
+/* Check if all slaves defined on the link can be found */
+static bool link_slaves_found(struct snd_sof_dev *sdev,
+			      struct snd_soc_acpi_link_adr *link,
+			      struct sdw_intel_ctx *sdw)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct sdw_intel_slave_id *ids = sdw->ids;
+	int num_slaves = sdw->num_slaves;
+	unsigned int part_id, link_id, unique_id, mfg_id;
+	int j, k;
+
+	/*
+	 * Spec definition
+	 *   Register           Bit     Contents
+	 *   DevId_0 [7:4]      47:44   sdw_version
+	 *   DevId_0 [3:0]      43:40   unique_id
+	 *   DevId_1            39:32   mfg_id [15:8]
+	 *   DevId_2            31:24   mfg_id [7:0]
+	 *   DevId_3            23:16   part_id [15:8]
+	 *   DevId_4            15:08   part_id [7:0]
+	 *   DevId_5            07:00   class_id
+	 */
+	for (k = 0; k < link->num_adr; k++) {
+		mfg_id = (link->adr[k] >> 24) & GENMASK(15, 0);
+		part_id = (link->adr[k] >> 8) & GENMASK(15, 0);
+		for (j = 0; j < num_slaves; j++) {
+			if (ids[j].id.part_id != part_id ||
+			    ids[j].id.mfg_id != mfg_id)
+				continue;
+			link_id = ids[j].link_id;
+			if (link->mask & BIT(link_id)) {
+				/*
+				 * we have to check unique id
+				 * if there are more than one
+				 * slave on the link
+				 */
+				unique_id = (link->adr[k] >> 40) &
+					    GENMASK(3, 0);
+				if (link->num_adr == 1 ||
```

Comment:
@bardliao what does num_adr == 1 check for? Its.not clear from the comment. U say if there is more than one slave but how do u know there is more than one slave?

Reply:
> @bardliao what does num_adr == 1 check for? Its.not clear from the comment. U say if there is more than one slave but how do u know there is more than one slave?

@ranj063 num_adr means how many slaves are presented on the adr array. So, we expect there is more than one slave on the link if num_adr > 1.